### PR TITLE
[NEW] 2 Factor Authentication when using OAuth and SAML

### LINKED ISSUE #4961: REST API - LDAP login not working
Your Rocket.Chat version: (make sure you are running the latest)

Hi,
I have LDAP set and user can properly login from the UI.
They are synchronized properly but when I try login in (first login) through the API I get 401 "Unauthorized"
Is there any special configuration that I am missing?

After the initial load from login via UI the API login works fine

Tks

### LINKED ISSUE #5499: Login with LDAP via Iframe
Hi,
I have a Rocket.Chat instance on my server with standalone LDAP server. Iframe.contentWindow.PostMessage with login-with-token event works great for logging users, but only for those previously logged in with Meteor.loginWithLDAP method. Is there a way to login users with LDAP via Iframe (or API) ? Or just run Meteor.loginWithLDAP from iframe parent window?

### LINKED ISSUE #6750: Require Two Factor Auth for external authenticators
Rocket.Chat Version: 0.55.1
Running Instances: 1
DB Replicaset OpLog: off
Node Version: 4.7.3

Hello Devs.

I have an issue with 2FA. After enabling it i can still login without entering a 2FA code. Is that a known issue?

thanks and cheers

### LINKED ISSUE #7642: API login doesn't work with an LDAP user when "Login Fallback" parameter is False
### Description:
Configure LDAP auth with "Login Fallback" parameter to False.

Log yourself with a valid user normally to validate that the user/pass is valid.

Use the same user/pass to perform an API authentication:
```
curl -H "Content-type:application/json" http://localhost:3000/api/v1/login -d '{ "username": "supervisor", "password": "xxxxxxx" }'
{
  "status": "error",
  "message": "Unauthorized"
}
```
Change the "Login Fallback" parameter to True, then log yourself again on the web interface with the same user to store the password in the local database, you then set the "Login Fallback" parameter to False again, it doesn't mater.

The authentication will now work with the same login/pass:
```
curl -H "Content-type:application/json" http://localhost:3000/api/v1/login -d '{ "username": "supervisor", "password": "xxxxxxx" }'
{
  "status": "success",
  "data": {
    "authToken": "KJQhGtdtBb0ecSn3CUQCpCWNUB6nVhpA226GEMM_72M",
    "userId": "tkBC5AEQvDLZ5SqNc"
  }
}
```
The API should ask LDAP when a user tries to use it and LDAP is active.
### Server Setup Information:

* Version of Rocket.Chat Server: 0.57.2
* Operating System: Ubuntu 16.04
* Deployment Method(snap/docker/tar/etc): tar
* Number of Running Instances: 1
* DB Replicaset Oplog: no
* Node Version: 4.5.0

### LINKED ISSUE #11411: Fail to login with  the REST API
I used the REST API to login, but the following error happened. I used wireshark to observe and although 200 is returned, not the json data, one html data is returned. I don't understand where the problem is.

 File "D:/06_PythonProject/Test/RocketChat.py", line 19, in <module>
    main()
  File "D:/06_PythonProject/Test/RocketChat.py", line 12, in main
    rocket = RocketChat('XXXX', 'XXXXX', server_url='https://open.rocket.chat/home')
  File "C:\Python27\lib\site-packages\rocketchat_API\rocketchat.py", line 21, in __init__
    self.login(user, password)
  File "C:\Python27\lib\site-packages\rocketchat_API\rocketchat.py", line 78, in login
    if login_request.json().get('status') == "success":
  File "C:\Python27\lib\site-packages\requests\models.py", line 896, in json
    return complexjson.loads(self.text, **kwargs)
  File "C:\Python27\lib\json\__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "C:\Python27\lib\json\decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Python27\lib\json\decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded

### LINKED ISSUE #11570: [BUG] 2FA Challenge is not provided with LDAP authentication
### Description:

When enabling 2FA with LDAP the Challenge isn't being provided and 2FA is just passed.

### Steps to reproduce:

1. Enable 2FA
2. Log in

### Expected behavior:

RocketChat asks for the 2FA token.

### Actual behavior:

RocketChat isn't asking for the 2FA token.

### Server Setup Information:

- Version of Rocket.Chat Server: 0.66.3
- Operating System: Ubuntu 16.04
- Deployment Method:  snap
- Number of Running Instances: 1
- DB Replicaset Oplog: enabled
- NodeJS Version: As shipped 
- MongoDB Version: As shipped

### Additional context

No additional context

### Relevant logs:

No relevant logs

### PR DESCRIPTION
**Pull request by Hudell**
Closes #9775
Closes #11731
Closes #6750
Closes #11570
Closes #5499
Closes #4961
Closes #7642
Closes #11411

This PR implements 2 Factor Authentication on all OAuth methods. Also includes 2FA for SAML and LDAP authentication.

### CODE REVIEW

File: packages/rocketchat-2fa/client/TOTPSaml.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	if (!options) {
+		options = {};
+		options.credentialToken = `id-${ Random.id() }`;
+	}
+	const { credentialToken } = options;
+
+	Accounts.saml.initiateLogin(options, function(/* error, result */) {
+		Accounts.callLoginMethod({
+			methodArguments: [{
+				saml: true,
```
Comment (rodrigok) [inline packages/rocketchat-2fa/client/TOTPSaml.js]: Isn't this inverted? `totp` property shouldn't be the top one?

Comment (Hudell) [reply packages/rocketchat-2fa/client/TOTPSaml.js]: All three properties are on the same object on the first index of the array, the property order doesn't matter.

Comment (rodrigok) [reply packages/rocketchat-2fa/client/TOTPSaml.js]: I wasn't talking about the order, but something like:
```
methodArguments: [{
	totp: {
		code,
		login: {
			saml: true,
			credentialToken,
		}
	},
}],
```

**Comment by rodrigok** [comment]
To support the new 2FA system with one-time login credentials (OAuth) needs to implement a new temporary mid-step authorization token (maybe a JWT) to proceed with the user's input of the 2FA code.

**Review by sampaiodiego — APPROVED** [review]

**Pull request merged**